### PR TITLE
export batch offline onnx model

### PR DESCRIPTION
**Pull request by Slyne**
Signed-off-by: slyned <slyned@nvidia.com>
1. Add codes for exporting onnx encoder & decoder.
2. Change offset from int to tensor for batching inference

There are warnings such as constant folding, which does not affect inference accuracy.
Not sure if speed/perf will be affected.

### CODE REVIEW

File: wenet/transformer/embedding.py [hunk 1]
```diff
@@ -41,28 +43,36 @@ def __init__(self,
...
             torch.Tensor: Encoded tensor. Its shape is (batch, time, ...)
             torch.Tensor: for compatibility to RelPositionalEncoding
         """
-        assert offset + x.size(1) < self.max_len
         self.pe = self.pe.to(x.device)
-        pos_emb = self.pe[:, offset:offset + x.size(1)]
+        if offset is None:
+            offset = torch.zeros(x.size(0), dtype=torch.int32).to(x.device)
+        index = offset.unsqueeze(1) + \
+            torch.arange(0, x.size(1)).to(offset.device)  # B X T
```
Comment (pengzhendong) [inline wenet/transformer/embedding.py:65]: X => x

File: wenet/transformer/embedding.py [hunk 2]
```diff
@@ -41,28 +43,36 @@ def __init__(self,
...
+        flag = index > 0
+        index = index * flag
+        pos_emb = F.embedding(index, self.pe)  # B X T X d_model
```
Comment (pengzhendong) [inline wenet/transformer/embedding.py:68]: X => x

File: wenet/transformer/embedding.py [hunk 3]
```diff
@@ -78,8 +88,13 @@ def position_encoding(self, offset: int, size: int) -> torch.Tensor:
         Returns:
             torch.Tensor: Corresponding encoding
         """
-        assert offset + size < self.max_len
-        return self.dropout(self.pe[:, offset:offset + size])
+        self.pe = self.pe.to(offset.device)
+        index = offset.unsqueeze(1) \
+            + torch.arange(0, size).to(offset.device)  # B X T
```
Comment (pengzhendong) [inline wenet/transformer/embedding.py:93]: X => x

File: wenet/transformer/embedding.py [hunk 4]
```diff
@@ -78,8 +88,13 @@ def position_encoding(self, offset: int, size: int) -> torch.Tensor:
...
+        flag = index > 0
+        index = index * flag
+        pos_emb = F.embedding(index, self.pe)  # B X T X d_model
```
Comment (pengzhendong) [inline wenet/transformer/embedding.py:96]: X => x

File: wenet/transformer/embedding.py [hunk 5]
```diff
@@ -90,42 +105,55 @@ class RelPositionalEncoding(PositionalEncoding):
...
             torch.Tensor: Encoded tensor (batch, time, `*`).
             torch.Tensor: Positional embedding tensor (1, time, `*`).
         """
-        assert offset + x.size(1) < self.max_len
+        if offset is None:
+            offset = torch.zeros(x.size(0), dtype=torch.int32).to(x.device)
+
         self.pe = self.pe.to(x.device)
+        index = offset.unsqueeze(1) \
+            + torch.arange(0, x.size(1)).to(offset.device)  # B X T
```
Comment (pengzhendong) [inline wenet/transformer/embedding.py:129]: X => x

File: wenet/transformer/embedding.py [hunk 6]
```diff
@@ -90,42 +105,55 @@ class RelPositionalEncoding(PositionalEncoding):
...
+        flag = index > 0
+        index = index * flag
+        pos_emb = F.embedding(index, self.pe)  # B X T X d_model
```
Comment (pengzhendong) [inline wenet/transformer/embedding.py:132]: X => x

**Review by pengzhendong — CHANGES_REQUESTED** [review]

**Review by pengzhendong — APPROVED** [review]

File: wenet/utils/mask.py [hunk 7]
```diff
@@ -145,7 +148,7 @@ def add_optional_chunk_mask(xs: torch.Tensor, masks: torch.Tensor,
     return chunk_masks
 
 
-def make_pad_mask(lengths: torch.Tensor) -> torch.Tensor:
+def make_pad_mask(lengths: torch.Tensor, max_len: int = None) -> torch.Tensor:
```
Comment (robin1001) [inline wenet/utils/mask.py:151]: why add the parameter here?

File: wenet/utils/mask.py [hunk 8]
```diff
@@ -35,8 +35,11 @@ def subsequent_mask(
          [1, 1, 0],
          [1, 1, 1]]
     """
-    ret = torch.ones(size, size, device=device, dtype=torch.bool)
-    return torch.tril(ret, out=ret)
+    arange = torch.arange(size, device=device)
```
Comment (robin1001) [inline wenet/utils/mask.py:38]: Add explanation about why change it like this.

File: wenet/transformer/subsampling.py [hunk 9]
```diff
@@ -16,7 +16,7 @@ def __init__(self):
         self.right_context = 0
         self.subsampling_rate = 1
 
-    def position_encoding(self, offset: int, size: int) -> torch.Tensor:
+    def position_encoding(self, offset: torch.Tensor, size: int) -> torch.Tensor:
```
Comment (robin1001) [inline wenet/transformer/subsampling.py:19]: We can directly use scalar type in onnx exporting  with opset_version>=11, please see https://pytorch.org/docs/stable/onnx.html

File: wenet/bin/export_onnx.py [hunk 10]
```diff
@@ -0,0 +1,263 @@
...
+
+    configs["encoder_conf"]["use_dynamic_chunk"] = False
+    model = init_asr_model(configs)
+    load_checkpoint(model, args.checkpoint)
+    model.eval()
+
+    bz = 32
+    seq_len = 100
+    beam_size = args.beam_size
+    feature_size = configs["collate_conf"]["feature_extraction_conf"]["mel_bins"]
```
Comment (robin1001) [inline wenet/bin/export_onnx.py:145]: better to use configs['input_dim'] instead of this.

File: requirements.txt [hunk 11]
```diff
@@ -6,3 +6,4 @@ tensorboard
 tensorboardX
 typeguard
 textgrid
+onnxruntime-gpu
```
Comment (robin1001) [inline requirements.txt:9]: Ask user to install it in export_onnx.py if the package is too big.

**Review by robin1001 — COMMENTED** [review]
please see inline.

File: wenet/utils/mask.py (hunk 7 above)
Comment (Slyne) [reply wenet/utils/mask.py:151]: It is for tensorrt export. Plus, giving pad mask the max actual length it needs to pad a batch will reduce the risk of the lengths tensor contains wrong maximum length.

File: wenet/utils/mask.py (hunk 8 above)
Comment (Slyne) [reply wenet/utils/mask.py:38]: > Add explanation about why change it like this.

Tril is not supported in onnx export. See issue: https://github.com/pytorch/pytorch/issues/34129

File: wenet/transformer/subsampling.py (hunk 9 above)
Comment (Slyne) [reply wenet/transformer/subsampling.py:19]: @robin1001  Yes, scalar is supportted. But for batching requests, we need to give each request an offset. It is ok for offline onnx model but not suitable for streaming.

File: wenet/bin/export_onnx.py (hunk 10 above)
Comment (Slyne) [reply wenet/bin/export_onnx.py:145]: Got it.

File: requirements.txt (hunk 11 above)
Comment (Slyne) [reply requirements.txt:9]: Got it.
